### PR TITLE
Don't force network-switching modal when wrong network is connected

### PR DESCRIPTION
**Pull request by peripheralist**
## What does this PR do and why?

Instead of forcing the connected wallet client to switch networks, display a "Wrong network" badge in the navbar.

Clicking the badge, or trying to create a transaction while on the wrong network, will display a bnc-onboard modal prompting the user to switch, OR will activate a network switch in the connected wallet client depending on the client.

## Screenshots or screen recordings

image

If using metamask, the native "switch network" modal will be triggered. If the wallet client doesn't have that function, this will show:
image

## Acceptance checklist

- [x] I have evaluated the [Approval Guidelines](https://github.com/jbx-protocol/juice-interface/blob/main/CONTRIBUTING.md#approval-guidelines) for this PR.
- [x] I have tested this PR in [all supported browsers](https://github.com/jbx-protocol/juice-interface/blob/main/CONTRIBUTING.md#supported-browsers).
- [x] I have tested this PR in dark mode and light mode (if applicable).

### CODE REVIEW

File: src/components/Navbar/Account.tsx [hunk 1]
```diff
@@ -1,16 +1,20 @@
...
 export default function Account() {
-  const { userAddress, signingProvider, onSelectWallet } =
+  const { userAddress, onSelectWallet, shouldSwitchNetwork, walletIsReady } =
     useContext(NetworkContext)
+  const {
+    theme: { colors },
+  } = useContext(ThemeContext)
 
-  if (!signingProvider) {
+  if (!userAddress) {
```
Comment (peripheralist) [inline src/components/Navbar/Account.tsx:17]: we still want to show the wallet address if one is connected

File: src/components/ReconfigurationStrategy/ReconfigurationStrategySelector.tsx [hunk 2]
```diff
@@ -23,14 +23,13 @@ function CustomStrategyInput({
...
 }) {
-  const { signerNetwork } = useContext(NetworkContext)
   return (
     <div>
       <Form.Item
         extra={
           <Trans>
             The address of any smart contract deployed on{' '}
-            {signerNetwork ?? 'mainnet'} that implements{' '}
+            {readNetwork.name ?? 'mainnet'} that implements{' '}
```
Comment (peripheralist) [inline src/components/ReconfigurationStrategy/ReconfigurationStrategySelector.tsx:32]: pretty sure this makes sense? not sure why we were using signerNetwork before

File: src/contexts/networkContext.ts [hunk 3]
```diff
@@ -1,13 +1,12 @@
...
-import { NetworkName } from 'models/network-name'
 import { createContext } from 'react'
 
 export const NetworkContext: React.Context<{
   signingProvider?: Web3Provider
-  signerNetwork?: NetworkName
   userAddress?: string
-  onNeedProvider?: () => Promise<void>
+  shouldSwitchNetwork?: boolean
+  walletIsReady?: () => Promise<boolean>
```
Comment (peripheralist) [inline src/contexts/networkContext.ts:9]: call this to ensure a wallet is connected & using the right network

File: src/pages/_app.page.tsx [hunk 4]
```diff
@@ -53,38 +33,6 @@ const AppWrapper: React.FC = ({ children }) => {
         <Navbar />
         <Content style={isMobile ? { paddingTop: 40 } : {}}>{children}</Content>
       </Layout>
-
-      <Modal
```
Comment (peripheralist) [inline src/pages/_app.page.tsx:57]: byeeee

File: src/providers/NetworkProvider.tsx [hunk 5]
```diff
@@ -19,9 +16,7 @@ export const NetworkProvider: React.FC = ({ children }) => {
   const { isDarkMode } = useContext(ThemeContext)
```
Comment (peripheralist) [inline src/providers/NetworkProvider.tsx:17]: some general reorganization here in addition to supporting new NetworkProvider properties.

main idea is replacing as much logic as possible with onboard subscriptions

File: src/providers/NetworkProvider.tsx [hunk 6]
```diff
@@ -109,24 +84,26 @@ export const NetworkProvider: React.FC = ({ children }) => {
...
-        signerNetwork: network,
-        signingProvider:
-          signingProvider && network === readNetwork.name && account
-            ? signingProvider
-            : undefined,
-        userAddress: account,
-        onSelectWallet: selectWallet,
-        onLogOut: logOut,
+        shouldSwitchNetwork: signingProvider && !usingCorrectNetwork,
+        signingProvider: usingCorrectNetwork ? signingProvider : undefined,
```
Comment (peripheralist) [inline src/providers/NetworkProvider.tsx:102]: we ONLY define NetworkContext.signingProvider if using correct network.

this may not be completely intuitive, but it's a simple way to ensure ContractReaders don't try reading the wrong network, and that no transactions are created to the wrong network

**Comment by peripheralist** [comment]
This is one of those PRs where testing cross-browser def matters. I'm having issues with my non-brave wallets (on prod as well), if someone else could test maybe... 🥹

File: src/contexts/networkContext.ts [hunk 7]
```diff
@@ -1,13 +1,12 @@
 import { Web3Provider } from '@ethersproject/providers'
 
-import { NetworkName } from 'models/network-name'
 import { createContext } from 'react'
 
 export const NetworkContext: React.Context<{
   signingProvider?: Web3Provider
-  signerNetwork?: NetworkName
   userAddress?: string
```
Comment (peripheralist) [inline src/contexts/networkContext.ts:7]: i'd _really_ like to change `userAddress` to `connectedWallet`. haven't done so to avoid scope creep, but can we plz?

**Comment by benschac** [comment]
This is dope.

I only had some UX comments/nits:
-  The `you must change networks` modal pops up after selecting a new network. In the gif I'm going from mainnet to rink and it'll flash for a second. Users might think they missed something important:

CleanShot 2022-08-04 at 17 22 23


- 
image
Personally like a lil more breathing room, but just like consistent spacing either way.


I only really skimmed the code, but it looked 💯.  This UX is sweet.

File: src/contexts/networkContext.ts (hunk 7 above)
Comment (benschac) [reply src/contexts/networkContext.ts:7]: Suggestion: Could do a stacked PR, on top of this PR that just changes the name of `userAddress` to `connectedWallet` so it's in a different scope and doesn't get mixed in with this feature work.

Also, not trying to push anything on anyone, I will stop making the suggestion after dis.

File: src/providers/NetworkProvider.tsx (hunk 6 above)
Comment (benschac) [reply src/providers/NetworkProvider.tsx:102]: > This is one of those PRs where testing cross-browser def matters. I'm having issues with my non-brave wallets (on prod as well), if someone else could test maybe... 🥹

I can test tomorrow if y'all still need it.

**Comment by benschac** [comment]
> This is one of those PRs where testing cross-browser def matters. I'm having issues with my non-brave wallets (on prod as well), if someone else could test maybe... 🥹

I can test tomorrow if y'all still need it.

**Comment by peripheralist** [comment]
> > This is one of those PRs where testing cross-browser def matters. I'm having issues with my non-brave wallets (on prod as well), if someone else could test maybe... 🥹
> 
> I can test tomorrow if y'all still need it.

i got it, ty ser 🙏

**Review by wraeth-eth — APPROVED** [review]

**Comment by peripheralist** [comment]
> image
> 
> Personally like a lil more breathing room, but just like consistent spacing either way.
> I only really skimmed the code, but it looked 💯. This UX is sweet.

good catch. the spacing is a little odd to get uniform since there's some custom styling on the language/theme selectors. i don't really mind the bunching since the wrong network tag relates directly to the wallet. gonna leave it for now and address the styling in another PR if needed
